drivers/dai/intel/alh/alh.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2022 Intel Corporation.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <errno.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <spinlock.h>
+#include <devicetree.h>
+#include <logging/log.h>
+
+#define DT_DRV_COMPAT intel_alh_dai
+#define LOG_DOMAIN dai_intel_alh
+
+LOG_MODULE_REGISTER(LOG_DOMAIN);
+
+#include "alh.h"
+
+/* Digital Audio interface formatting */
+static int dai_alh_set_config_tplg(struct dai_intel_alh *dp, const struct dai_config *common_config,
+				   const void *spec_config)
+{
+	struct dai_intel_alh_pdata *alh = dai_get_drvdata(dp);
+	const struct dai_intel_ipc3_alh_params *config = spec_config;
+
+	if (config->channels || config->rate) {
+		alh->params.channels = config->channels;
+		alh->params.rate = config->rate;
+		LOG_INF("%s channels %d rate %d", __func__, config->channels, config->rate);
+	}
+
+	alh->params.stream_id = config->stream_id;
+
+	return 0;
+}
+
+static int dai_alh_set_config_blob(struct dai_intel_alh *dp, const void *spec_config)
+{
```

Comment:
This again is mirroring the design in https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/dai/intel/ssp/ssp.c . We in middle of a transition with two ways to pass the configuration details and ALH has the same problem as SSP.
We have have ongoing discussion (with @lgirdwood , @juimonen and @abonislawski ) on how much app side wrapper/glue code we need for each driver. With this approach, we don't really need any glue code, but if we have the blob-to-different API conversion on app side, we need to a SW component on app side for each Zephyr driver, and this has been something we've wanted to avoid.

Reply:
Yeap, you are right, this doesnt looks great but it was upstreamed this way (dai/intel/ssp.c).

I think we need to change dai configuration exactly as you described for all dai drivers with one config struct and the rest in app layer but this needs to be done outside of ALH driver PR